src/lib/Microsoft.Health.Common/EventHub/Listeners/EventHubListener.cs

```diff
@@ -0,0 +1,498 @@
+        {
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            await _eventProcessorHost.RegisterEventProcessorFactoryAsync(this, _options.EventProcessorOptions);
+            _started = true;
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            if (_started)
+            {
+                await _eventProcessorHost.UnregisterEventProcessorAsync();
+            }
+
+            _started = false;
+        }
+
+        IEventProcessor IEventProcessorFactory.CreateEventProcessor(PartitionContext context)
+        {
+            return new EventProcessor(_options, _executor, _logger, _singleDispatch);
+        }
+
+        public IScaleMonitor GetMonitor()
+        {
+            return _scaleMonitor.Value;
+        }
+
+        /// <summary>
+        /// Wrapper for un-mockable checkpoint APIs to aid in unit testing
+        /// </summary>
+#pragma warning disable SA1201 // Elements should appear in the correct order
+        internal interface ICheckpointer
+#pragma warning restore SA1201 // Elements should appear in the correct order
+        {
+            Task CheckpointAsync(PartitionContext context);
+        }
+
+        // We get a new instance each time Start() is called.
+        // We'll get a listener per partition - so they can potentialy run in parallel even on a single machine.
+
+#pragma warning disable CA1303 // Do not pass literals as localized parameters
+        internal class EventProcessor : IEventProcessor, IDisposable, ICheckpointer
+        {
+            private readonly ITriggeredFunctionExecutor _executor;
+            private readonly bool _singleDispatch;
+            private readonly ILogger _logger;
+            private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+            private readonly ICheckpointer _checkpointer;
+            private readonly int _batchCheckpointFrequency;
+            private readonly int _eventQueueSize;
+            private readonly Task _timedFlush;
+            private readonly TimeSpan _flushTimeSpan;
+            private readonly ConcurrentDictionary<string, PartitionEventData> _pendingEventData;
+            private int _batchCounter = 0;
+            private bool _disposed = false;
+
+            public EventProcessor(EventHubOptions options, ITriggeredFunctionExecutor executor, ILogger logger, bool singleDispatch, ICheckpointer checkpointer = null)
+            {
+                _checkpointer = checkpointer ?? this;
+                _executor = executor;
+                _singleDispatch = singleDispatch;
+                _batchCheckpointFrequency = options.BatchCheckpointFrequency;
+                _logger = logger;
+                _pendingEventData = new ConcurrentDictionary<string, PartitionEventData>();
+                _flushTimeSpan = TimeSpan.FromMinutes(1);
+                _timedFlush = Task.Run(TimedFlush);

```

Comment:
In order to support time-based flushing, won't the function app need to running? And if its running, won't it be incurring cost?

Reply:
Even without time based flushing the function is always running since the underlying Event Hub processor needs to poll for Events.  It is probably to think of it better as a timed based buffer since in the normal mode events are immediately sent to the function (name in the code not withstanding).  This instead waits to the specified time to send the batch to reduce the calls on the FHIR server. This very helpful if you are using the SampledData type to collect many data points in the same observation.  If you are just using ValueQuantity or some other single use type it is less beneficial since each message will become an observation no matter what.

The scale to zero solution with KEDA works independently of this trigger.  In a KEDA scenario, KEDA would monitor the event hub and then spin up the first instance of the function.  At that point data would be buffered and sent to the FHIR service a regular intervals.